adoptingerlang.org

```diff
@@ -3444,44 +3446,74 @@ initContainers:
-
-Another use is to begin sending =503= status responses after receiving the =SIGTERM= signal from Kubernetes that tells the node to shutdown. It isn't required because Kubernetes will remove the =Pod= from the routes either way if it has sent it a shutdown signal, but this removal is not done before sending the signal, so for a period there will still be requests and readiness checks after the node has received the signal.
-
-The only necessary thing to do for graceful shutdown is to delay =init:stop()= from running until the =Pod= is removed from the =kube-proxy= =Endpoint=. The process can wait up to 30 seconds before Kubernetes sends a =SIGKILL= signal that will force the process to terminate.
-
-In =service_discovery= the =readinessProbe= is on a separate path, =/ready=, and with a higher frequency, every 5 seconds.
+**** Readiness, Liveness and Startup Probes
+
+Each container in a =Pod= can define a =readiness=, =liveness= and =startup= probe. The
+=liveness= probe . A startup probe can be added to delay when the =liveness= and
+=readiness= probes start being checked.
+
+Probes are defined either as a command to execute in the container, as an HTTP
+=GET= request made to a given port and path, or as an attempt to open a TCP
+connection to a specific port.
+
+A =livenessProbe= is not required to ensure restart of failed containers. If the
+process in the container crashes or stops for any reason, it will already be
+restarted. The probe should only be used for cases the process does not crash
+but for some reason has deadlocked and can not respond to requests. For this
+reason if a =livenessProbe= is included it should be kept extremely simple. It is
+more often than not safer to leave out the =livenessProbe= entirely and rely on
+the node crashes, metric based alerting and auto-scaling.
+
+One part of auto-scaling is based on the =readinessProbe=. The =readinessProbe=
+tells Kubernetes the =Pod= is ready to receive traffic. When a =Pod= becomes =unready=
+because a container is failing the =readinessProbe=, it is removed from the
+=Service= resulting in traffic being directed at the existing =Pods=, increasing
+their load, then a new =Pod= will be started.
+
+Not using the =livenessProbe= for this purpose means the containers are still
+available, since they aren't killed, and the reason for failing the check can be
+examined. This may mean attaching a shell if it isn't completely frozen or
+forcing a crash dump to be written. In a very resource constrained environment
+you may prefer to have the =livenessProbe= in order to have deadlocked containers
+and their =Pods= cleaned up right away, instead of waiting for manual intervention
+and the auto-scaler to scale down after load falls.
```

Comment:
A tricky pattern here is that if you use readiness probes to help you scale, and your auto-scaler is based on CPU usage, you can find yourself in a situation where during overload you want to refuse traffic, wall yourself off, and start gradually decreasing the resource pool you have, which in turn makes the problem worse.

Reply:
Do you mean a use-case of failing the readinessProbe during overload?

Reply:
It might be me who's wrong, but I thought the HPA only cared about readiness on the initial pod boot time, and not after that; so if down the road the pod sets itself as unready when it was and starts using less CPU, then the cluster's CPU average is going down and can prevent actual autoscaling.